File: Frontend/types/project.ts
/**
 * Project Management Types
 */

export interface ProjectMember {
  member_id: number;
  project_public_id: string;
  employee_public_id: string;
  employee_name: string;
  role_in_project: "Lead" | "Developer" | "Designer" | "QA" | "DevOps";
  joined_at: string;
}

export interface Project {
  public_id: string;
  project_code: string;
  project_name: string;
  description: string;
  status: "planning" | "active" | "on_hold" | "completed" | "cancelled";
  start_date: string;
  end_date?: string | null;
  head_employee_public_id?: string | null;
  head_employee_name?: string | null;
  project_head_public_id?: string | null;
  project_head_name?: string | null;
  members_count: number;
  member_count?: number;
  members?: ProjectMember[];
  completion_percentage?: number;
}

File: employee_management_backend_java/EmployeeManagment/EmployeeManagment/src/main/java/com/datansh/EmployeeManagment/entity/Project.java
package com.datansh.EmployeeManagment.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "projects", indexes = {
    @Index(name = "idx_projects_public_id", columnList = "public_id"),
    @Index(name = "idx_projects_head", columnList = "project_head_id"),
    @Index(name = "idx_projects_status", columnList = "status")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Project {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "project_id")
    private Long projectId;

    @Column(name = "public_id", nullable = false, unique = true, updatable = false)
    private UUID publicId;

    @Column(name = "project_name", nullable = false, length = 200)
    private String projectName;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_head_id")
    private Employee projectHead;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Column(name = "status", length = 20)
    @Builder.Default
    private String status = "planning"; // 'planning', 'active', 'on_hold', 'completed', 'cancelled'

    @OneToMany(mappedBy = "project", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    private List<ProjectMember> members = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    public void prePersist() {
        if (this.publicId == null) {
            this.publicId = UUID.randomUUID();
        }
        OffsetDateTime now = OffsetDateTime.now();
        if (this.createdAt == null) {
            this.createdAt = now;
        }
        this.updatedAt = now;
        if (this.status == null) {
            this.status = "planning";
        }
    }

    @PreUpdate
    public void preUpdate() {
        this.updatedAt = OffsetDateTime.now();
    }
}
